src/Sentry/MetricAggregator.cs

```diff
@@ -0,0 +1,459 @@
+                _captureCodeLocations(codeLocations);
+                _options.LogDebug("Code locations flushed: ", timestamp);
+            }
+
+            ClearStaleLocations();
+        }
+        catch (OperationCanceledException)
+        {
+            _options.LogInfo("Shutdown token triggered. Exiting metric aggregator.");
+        }
+        catch (Exception exception)
+        {
+            _options.LogError(exception, "Error processing metrics.");
+        }
+        finally
+        {
+            _flushLock.Release();
+        }
+    }
+
+    /// <summary>
+    /// Returns the keys for any buckets that are ready to be flushed (i.e. are for periods before the cutoff)
+    /// </summary>
+    /// <param name="force">Forces all buckets to be flushed, ignoring the cutoff</param>
+    /// <returns>
+    /// An enumerable containing the keys for any buckets that are ready to be flushed
+    /// </returns>
+    internal IEnumerable<long> GetFlushableBuckets(bool force = false)
+    {
+        if (!_buckets.IsValueCreated)
+        {
+            yield break;
+        }
+
+        if (force)
+        {
+            // Return all the buckets in this case
+            foreach (var key in Buckets.Keys)
+            {
+                yield return key;
+            }
+        }
+        else
+        {
+            var cutoff = MetricHelper.GetCutoff();
+            foreach (var key in Buckets.Keys)
+            {
+                var bucketTime = DateTimeOffset.FromUnixTimeSeconds(key);
+                if (bucketTime < cutoff)
+                {
+                    yield return key;
+                }
+            }
+        }
+    }
+
+    private Dictionary<long, Dictionary<MetricResourceIdentifier, SentryStackFrame>> FlushableLocations()
+    {
+        _codeLocationLock.Wait();
+        try
+        {
+            var result = _pendingLocations;
+            _pendingLocations = new Dictionary<long, Dictionary<MetricResourceIdentifier, SentryStackFrame>>();
+            return result;
+        }
+        finally
+        {
+            _codeLocationLock.Release();
+        }
+    }
+
+    /// <summary>
+    /// Clear out stale seen locations once a day
+    /// </summary>
+    private void ClearStaleLocations()
+    {
+        var now = DateTimeOffset.UtcNow;
+        var today = now.GetDayBucketKey();
+        if (lastClearedStaleLocations == today)
+        {
+            return;
+        }
+        // Allow 60 seconds for all code locations to be sent at the transition from one day to the next
+        const int staleGraceInMinutes = 1;
+        if (now.Minute < staleGraceInMinutes)
+        {
+            return;
+        }
+
+        foreach (var dailyValues in _seenLocations.Keys.ToArray())
+        {
+            if (dailyValues < today)
+            {
+                _seenLocations.TryRemove(dailyValues, out _);
+            }
+        }
+        lastClearedStaleLocations = today;
```

Comment:
is this a field? it's hard now with some fields prefixed with `_` and some not. Is that due to the new constructor syntax?

Reply:
No I think this just got overlooked. Maybe it wasn't private before or something... have changed.